frontend/javascripts/oxalis/view/right-border-tabs/segments_tab/segments_view.tsx

```diff
@@ -461,6 +505,67 @@ class SegmentsView extends React.Component<Props, State> {
     }
   }
 
+  onExpandTree = (expandedKeys: Key[]) => {
+    this.expandGroups(expandedKeys);
+  };
+
+  getSubGroupsAsTreeNodes = (groupId: number) => {
+    if (groupId !== MISSING_GROUP_ID) {
+      return getGroupByIdWithSubgroups(this.props.segmentGroups, groupId)
+        .filter((group) => group !== groupId)
+        .map((group) => getKeyForGroupId(group));
+    }
+    const allSegmentGroups = this.props.segmentGroups.flatMap((group) =>
+      getGroupByIdWithSubgroups(this.props.segmentGroups, group.groupId),
+    );
+    return allSegmentGroups.map((group) => getKeyForGroupId(group));
+  };
+
+  expandGroups = (newExpandedGroups: Key[]) => {
+    if (this.props.visibleSegmentationLayer == null) return;
+    const expandedKeySet = new Set(newExpandedGroups);
+    if (this.state.isRootGroupExpanded && !expandedKeySet.has(getKeyForGroupId(MISSING_GROUP_ID))) {
+      this.setState({ isRootGroupExpanded: false });
+    } else if (
+      !this.state.isRootGroupExpanded &&
+      expandedKeySet.has(getKeyForGroupId(MISSING_GROUP_ID))
+    ) {
+      this.setState({ isRootGroupExpanded: true });
+    }
+    const newGroups = mapGroups(this.props.segmentGroups, (group) => {
+      if (expandedKeySet.has(getKeyForGroupId(group.groupId))) {
+        return {
+          ...group,
+          isExpanded: true,
+        };
+      } else {
+        return {
+          ...group,
+          isExpanded: false,
+          // Close all groups that are not in the expanded list so this method
+          // can be called for every update, e.g. when a group is collapsed.
+        };
```

Comment:
this way, you don't create new objects for all groups where the isExpanded property does not change:

```suggestion
      const isExpanded = expandedKeySet.has(getKeyForGroupId(group.groupId));
      if (group.isExpanded === isExpanded) {
        return group;
      } else {
        return {
          ...group,
          isExpanded,
        };
      }
```